### ISSUE TITLE
Further Enhancements

### ISSUE THREAD

**Issue by gXLg** [gXLg/libr-getter#23]
* ~Add version update checker~ (done in [v2.0.0](https://github.com/gXLg/libr-getter/releases/tag/v2.0.0)) ~(potentially auto-updater)~
* ~Upload to modrinth~ ~worklflow is set up, waiting for Modrinth to approve the project~ done (#21)
* ~Add safe-checker: check if when lectern broken, the villager can escape~ (done in [v2.1.1](https://github.com/gXLg/libr-getter/releases/tag/v2.1.1))
* ~Potentially: Work on a GUI~ (done in in [v2.1.0](https://github.com/gXLg/libr-getter/releases/tag/v2.1.0))
* Maybe (if not too hard): work on path finder to pick up lectern (#15)
* Auto stop/auto eat when hungry, takes damage, or the villager is attacked (will not be wanted by Modrinth since it could be used as a hack)
* ~Add keybinds~ (done in [v2.1.0](https://github.com/gXLg/libr-getter/releases/tag/v2.1.0))
* ~Add translation~(done in [v2.0.2](https://github.com/gXLg/libr-getter/releases/tag/v2.0.2))
* ~Investigation needed: all-in-one mod, to use a single mod with different Minecraft versions (every feature I add, has to be added in other mod versions and can be very annoying to implement)~ (impossible? no! done in [v2.0.0](https://github.com/gXLg/libr-getter/releases/tag/v2.0.0))
* Add Mod Menu compatibility
* ~Add enchantments presets~ (wontdo)
* ~Add timeout replacing (#30)~ (done in [v2.2.0](https://github.com/gXLg/libr-getter/releases/tag/v2.2.0))
* Add tradehall automation

Users are welcome to add their ideas below

**Comment by MilesTFox** [comment]
Support for non-minecraft:NAME enchantments like enchantmentsolution:soulbound, etc.

**Comment by gXLg** [comment]
I was successfully able to merge CherryBlossom and Warden releases locally, going to work on merging further versions

**Comment by gXLg** [comment]
> * offhanding the lecterns, so you can place/break without switching slots.
> 
>     * non-librarian villagers/workstations, with ID matching. So like, look for the cheapest carrots on a farmer.
> 
>     * non-minecraft:NAME enchantments like enchantmentsolution:soulbound, etc.
> 
>     * non 'minecraft:enchanted_book' enchantments. enchantmentsolution ones are on minecraft:book

Are you able to add those `non-minecraft:NAME` enchantments through `/librget add <enchantment>`?

**Comment by MilesTFox** [comment]
I was not able to.

**Comment by gXLg** [comment]
That's bad, I will have to figure out, how to get all registered enchantments

**Comment by MilesTFox** [comment]
Might have to be a config file or allowing setting anything, because the
additional enchantments are server-side via a plugin, and there is no
client-side mod.

**Comment by gXLg** [comment]
Ah, that makes sense! Currently my mod parses only the data saved on the client. My data structure however would theoretically allow to enter "setting anything" as an option, so I will work on it and release in the next update! Thanks!

**Comment by gXLg** [comment]
@MilesTFox 
![image](https://github.com/gXLg/libr-getter/assets/65429873/067a46a0-1d51-457f-a184-accfd6afeb49)
This was actually quite doable. The update will soon come out!

**Comment by gXLg** [comment]
@MilesTFox added in [v2.0.1](https://github.com/gXLg/libr-getter/releases/tag/v2.0.1)

**Comment by SamAurel** [comment]
Could you recolor the remove button in the list (to cyan or something) and maybe add such a button to the success message (Successfully found whatever 2 after 45 tries for a price of 21 emeralds (remove))

**Comment by SamAurel** [comment]
Thanks for the fast response,
I used that for a bit, but stopped because I wanted more control over the goals list. Its not a pressing issue as I can just remove the goal with a quick command, just an idea I had. :)

**Comment by gXLg** [comment]
Alright, I think I might add it anyways

**Comment by gXLg** [comment]
@SamAurel the "remove" button is now there, if the config `removegoal` is deactivated

**Comment by ToanAnh312** [comment]
instead of adding path finding (which is more complex), consider having a blueprint that has water push to your direction instead :D 
![image](https://github.com/user-attachments/assets/cb6c1526-6322-4fbe-8d0a-5a2322ae3f1a)

**Comment by ToanAnh312** [comment]
Can i add Vietnamese :D

**Comment by gXLg** [comment]
@ToanAnh312 Yes, feel free to add it. You will have to add the translаtion file inside `src/main/resources/assets/librgetter/lang`. If you don't know how to do a pull request, you can drop the json file here

**Comment by ToanAnh312** [comment]
@gXLg done, you can see it [here](https://github.com/gXLg/libr-getter/pull/32)

**Comment by gXLg** [comment]
@ToanAnh312 I have a new update coming today, it adds a new config.

In English it will be "Time in seconds to wait for a villager to update its profession before replacing the lectern. Set to 0 to disable."

Can you help me with a translation into Vietnamese of this one?

**Comment by ToanAnh312** [comment]
Here @gXLg "Thới gian (tính bằng giây) để đợi dân làng cập nhật nghề nghiệp trước khi thay bục đặt sách. Đặt thành 0 để tắt" :p

**Comment by ToanAnh312** [comment]
Also in google translate, you can see that Lectern mean "bục giảng", it is incorrect because it mean podium, another one is "giá để kinh" (desk) that is wrong as we use "giá" for sthing that hanging and "bục" for sthing has platform (eg:tripod)

**Comment by ZettaBytte** [comment]
Hi, it may be possible to add compatibility with Trade Cycle? Here is the link to the Modrinth page: https://modrinth.com/mod/trade-cycling

**Comment by gXLg** [comment]
@ZettaBytte, welcome! What kind of "compatibility" are you looking for? Do you want, that LibrGetter in combination with Trade Cycle would click the cycle button through, until the wished enchantment is found?

**Comment by ZettaBytte** [comment]
Exactly that is what I was thinking. I've been using LibrGetter in a server and noticed that a lot of the searching time was being taken by the villager not re-taking the job due to lag or even checking to get a job on other tables. Trade Cycle allows to refresh the trade without breaking the table, so if LibrGetter could use that button it would speed up the process both on multiplayer and singleplayer.

**Comment by gXLg** [comment]
@ZettaBytte alright, thanks for the idea! I will implement it in near future after the 1.21.5 drop.

**Comment by gXLg** [comment]
@ZettaBytte the compatibility with Trade Cycle has now been implemented, check the newest release and read the description for more details

**Comment by ToanAnh312** [comment]
The ring sound when found enchantment kinda short and silent, i hope we can add customizable sound effect

**Comment by gXLg** [comment]
@ToanAnh312 the mod already has many configs for the actual process, by adding new configs for other things it will make the configuration confusing and non-obvious. I may come back to it if I ever implement config categories, but until then it doesn't look good and is in my opinion "off-topic" from the actual goal of the mod

**Comment by ToanAnh312** [comment]
@gXLg i think the way mod check the book somewhat not natural, can we add head motion or sth like that ?

**Comment by ToanAnh312** [comment]
Btw if we cant add custom sound effect, can you point me where do you store that sound ?

**Comment by gXLg** [comment]
@ToanAnh312 

> [@gXLg](https://github.com/gXLg) i think the way mod check the book somewhat not natural, can we add head motion or sth like that ?

I was contacted abt it on Discord, this will be implemented in the next update, when 1.21.6 drops.

> Btw if we cant add custom sound effect, can you point me where do you store that sound ?

It's the vanilla "Level Up" sound played at 0.7 speed.
This is the corresponding line in the code: [link](https://github.com/gXLg/libr-getter/blob/09021580d188b59a541d41da8214b07213bbb679/src/main/java/com/gxlg/librgetter/utils/reflection/Minecraft.java#L323)

**Comment by DOTLeo0913** [comment]
When I play single player or add a mod to the server, I would like the whole process of destroying the Lectern and checking the villagers to be done on the server side, not on the client.
If this can be done, we can speed up transactions with carpet mod tick rate.

**Comment by gXLg** [comment]
@DOTLeo0913 Hello, thanks for the suggestion, however this will not be implemented. The mod is intended to be client-side only.
I will see if I can sync client-side ticks with the server, this way there would be no need to install the mod on the server.

**Comment by ToanAnh312** [comment]
> @gXLg Seconded on that, base functionality needs to be kept compatible with vanilla servers.
> 
> 

Correct me if im wrong, i think the mod use vanilla code, just like sodium mod so it suppose to run flawless on any server
